### PR TITLE
argmax for half datatype fix

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28706 argmax for half datatype fix**

Differential Revision: [D18145687](https://our.internmc.facebook.com/intern/diff/D18145687)

### CODE REVIEW

**Review by peterbell10 — APPROVED** [review]
Looks good, just a minor comment.

File: aten/src/THC/THCDeviceUtils.cuh [hunk 1]
```diff
@@ -115,6 +116,17 @@ __device__ __forceinline__ T WARP_SHFL_DOWN(T value, unsigned int delta, int wid
     return __shfl_down(value, delta, width);
 #endif
 }
+template <>
+__device__ __forceinline__ c10::Half WARP_SHFL_DOWN<c10::Half>(c10::Half value, unsigned int delta, int width, unsigned int mask)
```
Comment (peterbell10) [inline aten/src/THC/THCDeviceUtils.cuh]: Why is this a template specialisation and not just a function overload?

**Review by ngimel — CHANGES_REQUESTED** [review]
Rather than adding `is_nan` and `warp_shfl_down` for halfs, I'd prefer to convert halfs to float on read, and do all the operations in float, like it's done for other reductions. While it does not matter for max operation, adding `shfl_down` for halfs will make it easy to overlook incorrect reduction unexpectedly truncating intermediate values to half in cases where it does matter (sums, products, norms etc). The implementation would look similar to other reductions, I checked that it compiles, but feel free to also template `*_impl` on the comparison functor:
```
template <typename scalar_t, typename acc_t=scalar_t>
void argmax_kernel_cuda_impl(TensorIterator& iter) {
  gpu_reduce_kernel<scalar_t, int64_t>(
    iter, 
    ArgMaxOps<acc_t>{},
    thrust::pair<scalar_t, int64_t>(at::numeric_limits<scalar_t>::lower_bound(), 0));
};
  
template <typename scalar_t, typename acc_t=scalar_t>
void argmin_kernel_cuda_impl(TensorIterator& iter) {
  gpu_reduce_kernel<scalar_t, int64_t>(
    iter, 
    ArgMinOps<acc_t>{},
    thrust::pair<scalar_t, int64_t>(at::numeric_limits<scalar_t>::lower_bound(), 0));
};

void argmax_kernel_cuda(TensorIterator& iter) {
  if (iter.dtype() == kHalf) {
    return argmax_kernel_cuda_impl<at::Half, float>(iter);
  } 
  AT_DISPATCH_ALL_TYPES(iter.dtype(), "argmax_cuda", [&]() {
    argmax_kernel_cuda_impl<scalar_t>(iter);
  });
}

void argmin_kernel_cuda(TensorIterator& iter) {
  if (iter.dtype() == kHalf) {
    return argmin_kernel_cuda_impl<at::Half, float>(iter);
  } 
  AT_DISPATCH_ALL_TYPES(iter.dtype(), "argmin_cuda", [&]() {
    argmin_kernel_cuda_impl<scalar_t>(iter);
  });
}
```

**Comment by ngimel** [comment]
I've also noticed, while we are at it, that `min_values/max_values` are not dispatched for half. It seems those functions are only accessible from C++, and I don't know how widely they are used in C++ frontend if at all, but probably it would make sense to fix that also?

**Comment by gchanan** [comment]
> I've also noticed, while we are at it, that min_values/max_values are not dispatched for half. It seems those functions are only accessible from C++, and I don't know how widely they are used in C++ frontend if at all, but probably it would make sense to fix that also?

Yes, it would make sense to fix these.  @pbelevich can you file an issue?

**Comment by gchanan** [comment]
> Rather than adding is_nan and warp_shfl_down for halfs, I'd prefer to convert halfs to float on read, and do all the operations in float, like it's done for other reductions. While it does not matter for max operation, adding shfl_down for halfs will make it easy to overlook incorrect reduction unexpectedly truncating intermediate values to half in cases where it does matter (sums, products, norms etc). The implementation would look similar to other reductions, I checked that it compiles, but feel free to also template *_impl on the comparison functor:

@ngimel: I'd like to get this fix in ASAP since this is a regression from our previous release.  Your changes look pretty easy to apply (correct me if I'm wrong).  Otherwise we can just do something like go back to the old implementation where we ran `max` for `argmax`.

**Comment by ngimel** [comment]
Yes, they are easy to apply, I've just basically posted the full (untested) diff, only one file is affected.  Do you want me to open a new PR, or can @pbelevich do it?

**Comment by pbelevich** [comment]
https://github.com/pytorch/pytorch/pull/28787

**Pull request closed without merging**